cpp/src/parquet/bloom_filter_benchmark.cc

```diff
@@ -0,0 +1,230 @@
+      data[i].ptr = heap->data() + i * kGenerateBenchmarkDataStringLength;
+      data[i].len = kGenerateBenchmarkDataStringLength;
+    }
+  } else if constexpr (std::is_same_v<Int96, T>) {
+    std::default_random_engine gen(/*seed*/ 0);
+    std::uniform_int_distribution<int> d(std::numeric_limits<int>::min(),
+                                         std::numeric_limits<int>::max());
+    for (uint32_t i = 0; i < size; ++i) {
+      data[i].value[0] = d(gen);
+      data[i].value[1] = d(gen);
+      data[i].value[2] = d(gen);
+    }
+  }
+}
+
+template <typename DType>
+static void BM_ComputeHash(::benchmark::State& state) {
+  using T = typename DType::c_type;
+  std::vector<T> values(kBloomFilterElementSize);
+  std::vector<uint8_t> heap;
+  GenerateBenchmarkData(kBloomFilterElementSize, values.data(), &heap);
+  auto filter = createBloomFilter(kBloomFilterElementSize);
+  for (auto _ : state) {
+    for (const auto& value : values) {
+      uint64_t hash = 0;
+      if constexpr (std::is_same_v<DType, FLBAType>) {
+        hash = filter->Hash(&value, kGenerateBenchmarkDataStringLength);
+      } else if constexpr (std::is_same_v<DType, Int96Type>) {
+        hash = filter->Hash(&value);
+      } else if constexpr (std::is_same_v<DType, ByteArrayType>) {
+        hash = filter->Hash(&value);
+      } else {
+        hash = filter->Hash(value);
+      }
+      ::benchmark::DoNotOptimize(hash);
+      ::benchmark::ClobberMemory();
+    }
+  }
+  state.SetItemsProcessed(state.iterations() * values.size());
+}
+
+template <typename DType>
+static void BM_BatchComputeHash(::benchmark::State& state) {
+  using T = typename DType::c_type;
+  std::vector<T> values(kBloomFilterElementSize);
+  std::vector<uint8_t> heap;
+  GenerateBenchmarkData(kBloomFilterElementSize, values.data(), &heap);
+  auto filter = createBloomFilter(kBloomFilterElementSize);
+  std::vector<uint64_t> hashes(kBloomFilterElementSize);
+  for (auto _ : state) {
+    if constexpr (std::is_same_v<DType, FLBAType>) {
+      filter->Hashes(values.data(), kGenerateBenchmarkDataStringLength,
+                     static_cast<int>(values.size()), hashes.data());
+    } else {
+      filter->Hashes(values.data(), static_cast<int>(values.size()), hashes.data());
+    }
+    ::benchmark::DoNotOptimize(hashes);
+    ::benchmark::ClobberMemory();
+  }
+  state.SetItemsProcessed(state.iterations() * values.size());
+}
+
+static void BM_InsertHash(::benchmark::State& state) {
+  using T = int32_t;
+  std::vector<T> values(kBloomFilterElementSize);
+  std::vector<uint8_t> heap;
+  GenerateBenchmarkData(kBloomFilterElementSize, values.data(), &heap);
+  auto filter = createBloomFilter(kBloomFilterElementSize);
+  std::vector<uint64_t> hashes(1024);
+  filter->Hashes(values.data(), static_cast<int>(values.size()), hashes.data());
+  for (auto _ : state) {
+    for (auto hash : hashes) {
+      filter->InsertHash(hash);
+      ::benchmark::ClobberMemory();
+    }
+  }
```

Comment:
Several things:
* I think we should create a new bloom filter for each benchmark iteration
* `ClobberMemory` should be moved after the inner loop, as it acts as a memory barrier and could pessimize performance here

Reply:
Okay, Seems that I should create BF and then start the clock like:

```c++
  for (auto _ : state) {
    state.PauseTiming();
    data = CreateBloomFilterAndInsertData();
    state.ResumeTiming();
    for (int j = 0; j < state.range(1); ++j)
      ...
  }
```

?